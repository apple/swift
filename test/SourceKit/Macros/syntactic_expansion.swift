//--- test.swift
@DelegatedConformance
@wrapAllProperties
struct Generic<Element> {

  @myPropertyWrapper
  @otherAttr
  var value: Int

  func member() {}
  var otherVal: Int = 1

  #bitwidthNumberedStructs("blah")
}

//--- DelegatedConformance.json
{
  key.macro_roles: [source.lang.swift.macrorole.conformance, source.lang.swift.macrorole.member],
  key.modulename: "MacroDefinition",
  key.typename: "DelegatedConformanceMacro",
}

//--- myPropertyWrapper.json
{
  key.macro_roles: [source.lang.swift.macrorole.accessor, source.lang.swift.macrorole.peer],
  key.modulename: "MacroDefinition",
  key.typename: "PropertyWrapperMacro",
}

//--- wrapAllProperties.json
{
  key.macro_roles: [source.lang.swift.macrorole.memberattribute],
  key.modulename: "MacroDefinition",
  key.typename: "WrapAllProperties",
}

//--- bitwidthNumberedStructs.json
{
  key.macro_roles: [source.lang.swift.macrorole.declaration],
  key.modulename: "MacroDefinition",
  key.typename: "DefineBitwidthNumberedStructsMacro",
}

//--- dummy.script
// REQUIRES: swift_swift_parser

// RUN: %empty-directory(%t)
// RUN: mkdir -p %t/plugins
// RUN: split-file %s %t

//##-- Prepare the macro plugin.
// RUN: %host-build-swift -swift-version 5 -emit-library -o %t/plugins/%target-library-name(MacroDefinition) -module-name=MacroDefinition %S/../../Macros/Inputs/syntax_macro_definitions.swift -g -no-toolchain-stdlib-rpath

// RUN: %sourcekitd-test -req=syntactic-expandmacro \
// RUN:   -req-opts=1:1:%t/DelegatedConformance.json \
// RUN:   -req-opts=5:3:%t/myPropertyWrapper.json \
// RUN:   -req-opts=2:1:%t/wrapAllProperties.json \
// RUN:   -req-opts=12:3:%t/bitwidthNumberedStructs.json \
// RUN:   %t/test.swift \
// RUN:   -- \
// RUN:   %t/test.swift \
// RUN:   -plugin-path %t/plugins -Xfrontend -dump-macro-expansions \
// RUN:   -module-name TestModule \
// RUN:   | tee %t.response

// RUN: diff -u %s.expected %t.response
