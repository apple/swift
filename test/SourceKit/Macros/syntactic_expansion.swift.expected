{
  key.categorizededits: [
    {
      key.edits: [
        {
          key.line: 1,
          key.column: 1,
          key.endline: 1,
          key.endcolumn: 22,
          key.text: ""
        }
      ],
      key.category: source.edit.kind.active
    },
    {
      key.edits: [
        {
          key.line: 13,
          key.column: 1,
          key.endline: 13,
          key.endcolumn: 1,
          key.text: "static func requirement() where Element : P {\n  Element.requirement()\n}"
        }
      ],
      key.category: source.edit.kind.active
    },
    {
      key.edits: [
        {
          key.line: 13,
          key.column: 2,
          key.endline: 13,
          key.endcolumn: 2,
          key.text: "extension Generic : P where Element: P {}"
        }
      ],
      key.category: source.edit.kind.active
    },
    {
      key.edits: [
        {
          key.line: 5,
          key.column: 3,
          key.endline: 5,
          key.endcolumn: 21,
          key.text: ""
        }
      ],
      key.category: source.edit.kind.active
    },
    {
      key.edits: [
        {
          key.line: 7,
          key.column: 17,
          key.endline: 7,
          key.endcolumn: 17,
          key.text: "{\n    get {\n            _value.wrappedValue\n      }\n\n    set {\n            _value.wrappedValue = newValue\n      }\n}"
        }
      ],
      key.category: source.edit.kind.active
    },
    {
      key.edits: [
        {
          key.line: 7,
          key.column: 12,
          key.endline: 7,
          key.endcolumn: 12,
          key.text: "var _value: MyWrapperThingy<Int>"
        }
      ],
      key.category: source.edit.kind.active
    },
    {
      key.edits: [
        {
          key.line: 2,
          key.column: 1,
          key.endline: 2,
          key.endcolumn: 19,
          key.text: ""
        }
      ],
      key.category: source.edit.kind.active
    },
    {
      key.edits: [
        {
          key.line: 7,
          key.column: 3,
          key.endline: 7,
          key.endcolumn: 3,
          key.text: "@Wrapper"
        }
      ],
      key.category: source.edit.kind.active
    },
    {
      key.edits: [
        {
          key.line: 10,
          key.column: 3,
          key.endline: 10,
          key.endcolumn: 3,
          key.text: "@Wrapper"
        }
      ],
      key.category: source.edit.kind.active
    },
    {
      key.edits: [
        {
          key.line: 12,
          key.column: 3,
          key.endline: 12,
          key.endcolumn: 35,
          key.text: "struct blah8 {\n  func macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu_() {\n  }\n  func macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu0_() {\n  }\n}\nstruct blah16 {\n  func macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu1_() {\n  }\n  func macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu2_() {\n  }\n}\nstruct blah32 {\n  func macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu3_() {\n  }\n  func macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu4_() {\n  }\n}\nstruct blah64 {\n  func macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu5_() {\n  }\n  func macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu6_() {\n  }\n}"
        }
      ],
      key.category: source.edit.kind.active
    }
  ]
}
